### PR TITLE
wandb.Audio

### PR DESCRIPTION
**Pull request by altay**
Client support for audio files!

To upload a file, you can do wandb.Audio(np_array, sample_rate=sample_rate, caption=caption).  

Or I can send you a test script if you want.

### CODE REVIEW

**Comment by shawnlewis** [comment]
Looks like this is failing on CI due to missing packages?

**Comment by shawnlewis** [comment]
I think you would in setup.py, but in many places we prefer not to have a hard dependency, and instead print an error if a user tries to use that feature without the required library (to minimize the amount of actual deps the library has). There should be an example of that in the code somewhere, I can look around in a bit.

**Comment by adrianbg** [comment]
If you don't want to have a hard dependency on a particular module, you can use @vanpelt's `util.get_module()` or just do a `try:/except ImportError: pass`

File: wandb/data_types.py [hunk 1]
```diff
@@ -30,18 +30,22 @@ def val_to_json(key, val, mode="summary", step=None):
...
     elif util.is_plotly_typename(typename):
         converted = util.convert_plots(val)
-    if isinstance(val, Image):
+    if isinstance(val, Image) or isinstance(val, Audio):
         val = [val]
 
     if isinstance(val, collections.Sequence) and len(val) > 0:
         is_image = [isinstance(v, Image) for v in val]
-        if all(is_image):
+        is_audio = [isinstance(v, Audio) for v in val]
```
Comment (adrianbg) [inline wandb/data_types.py]: Are None's allowed / possible here? Are they considered Images / Audios?

File: wandb/data_types.py [hunk 2]
```diff
@@ -663,6 +667,52 @@ def transform(table):
...
+        if sample_rates:
+            meta["sampleRates"] = sample_rates
+        captions = Audio.captions(audio_list[:Audio.MAX_AUDIO_COUNT])
+        if captions:
+            meta["captions"] = captions
+        return meta
+
+    @staticmethod
+    def sample_rates(audio_list):
+        if audio_list[0].sample_rate != None:
```
Comment (adrianbg) [inline wandb/data_types.py]: This seems a little weird. Why do you only care about the first Audio in the list? Why not just always return the list of all sample rates?

File: wandb/data_types.py [hunk 3]
```diff
@@ -663,6 +667,52 @@ def transform(table):
...
+            return [a.sample_rate for a in audio_list]
+        else:
+            return False
+
+    @staticmethod
+    def captions(audio_list):
+        if audio_list[0].caption != None:
```
Comment (adrianbg) [inline wandb/data_types.py]: Ditto here

File: wandb/data_types.py (hunk 1 above)
Comment (altay) [reply wandb/data_types.py]: If any is None, it throws the error "Mixed media types in the same list aren't supported."  

@shawnlewis , @vanpelt  Is that the desired behavior, or should we filter out None from that array?

File: wandb/data_types.py (hunk 2 above)
Comment (altay) [reply wandb/data_types.py]: @adrianbg Is this better?  https://github.com/wandb/client/pull/248/commits/42e2cc81fec02dcf49fdc2c96353d0f7515cd2e1

**Comment by altay** [comment]
@adrianbg Better now?

File: wandb/data_types.py (hunk 1 above)
Comment (shawnlewis) [reply wandb/data_types.py]: I'm not sure... can you think of a use case where you'd want to send None?

Comment (adrianbg) [reply wandb/data_types.py]: In a History where not all time steps have Images

**Pull request merged**